### PR TITLE
update lambda execution policy

### PR DESCRIPTION
**Pull request by clarisse-lau**
Fix permissive lambda execution policy

### CODE REVIEW

File: template.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ Resources:
             - s3:GetObject
             - s3:PutObject
             - s3:PutObjectAcl
-          Resource: arn:aws:s3:::*
+          Resource: arn:aws:s3:::htan-synapse-sync-*/*
```
Comment (zaro0508) [inline template.yaml:30]: This lambda is now setup in a way that's pretty generic. Do you think we could use this lambda for other projects, other than HTAN? maybe you want to be more generic with the prefix?   

Also how about we parameterize this to let the user that deploys this lambda decide the prefix name? for instance:
```
Resource: !Sub "arn:aws:s3:::${BucketNamePrefix}/*"
```

Comment (clarisse-lau) [reply template.yaml]: Sounds good to me. Made this change

File: template.yaml [hunk 2]
```diff
@@ -27,7 +30,7 @@ Resources:
             - s3:GetObject
             - s3:PutObject
             - s3:PutObjectAcl
-          Resource: arn:aws:s3:::*
+          Resource: !Sub "arn:aws:s3:::${BucketNamePrefix}*/*"
```
Comment (zaro0508) [inline template.yaml:33]: i think you can leave out the first `*` and let the user pass that in (i.e. my-bucket-*).

Comment (clarisse-lau) [reply template.yaml:33]: Ah, got it

File: template.yaml [hunk 3]
```diff
@@ -10,6 +10,11 @@ Parameters:
   KmsDecryptPolicyArn:
     Description: 'The KMS key policy ARN to access the Synapse service user secrets'
     Type: String
+  BucketNamePrefix:
+    Description: 'Prefix of buckets accessed by the lambda'
+    Type: String
+    ConstraintDescription: >-
+      Must be prefix of bucket name followed by wildcard (e.g. htan-dcc-*)
```
Comment (zaro0508) [inline template.yaml:17]: is wildcard really a requirement?  what if someone wants to use lambda with one specific bucket?

File: README.md [hunk 4]
```diff
@@ -176,16 +138,54 @@ hooks:
...
+  S3SynapseSyncFunctionRoleArn: !stack_output_external "s3-synapse-sync::FunctionRoleArn"
+
+# Due to circular dependencies, enabling bucket notification must be done after bucket creation"
+# https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig.html
+sceptre_user_data:
+  EnableNotificationConfiguration: "false"
+```
+
+Deploy with sceptre, Notification configuration is disabled on 1st deploy.
+Deploy a 2nd time with `EnableNotificationConfiguration: "true"`
```
Comment (zaro0508) [inline README.md:184]: we might be able to do this in one step if you are interested. just requires a little more trickery.

File: README.md [hunk 5]
```diff
@@ -176,16 +138,54 @@ hooks:
 parameters:
   BucketVariables: !file_contents "data/s3-synapse-sync-bucket-vars.yaml"
   KmsDecryptPolicyArn: !stack_output_external "s3-synapse-sync-kms-key::KmsDecryptPolicyArn"
+  BucketNamePrefix: "htan-dcc-*"
```
Comment (zaro0508) [inline README.md:141]: how about a corresponding PR for on the deployment side?  https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/config/prod/s3-synapse-sync.yaml#L14

File: template.yaml (hunk 3 above)
Comment (clarisse-lau) [reply template.yaml]: updated to remove this requirement

File: README.md (hunk 4 above)
Comment (clarisse-lau) [reply README.md:184]: I think unless we run into issues, this two-step deployment is ok

File: README.md (hunk 5 above)
Comment (clarisse-lau) [reply README.md:141]: PR for deployment side: https://github.com/Sage-Bionetworks/scicomp-provisioner/pull/384

**Pull request merged**